File: example/src/main/java/com/example/service/impl/ExampleServiceImpl.java
package com.example.service.impl;

import com.example.entity.AmlWhiteListEntity;
import com.zipe.base.annotation.DS;
import com.zipe.enums.ResourceEnum;
import com.example.entity.InfoEntity;
import com.example.entity.UserEntity;
import com.example.jdbc.ExampleJdbc;
import com.example.model.User;
import com.example.repository.AmlWhiteListRepository;
import com.example.repository.InfoRepository;
import com.example.repository.UserRepository;
import com.example.service.ExampleService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Slf4j
@Service
public class ExampleServiceImpl implements ExampleService {

    private final UserRepository userRepository;
    private final InfoRepository infoRepository;
    private final AmlWhiteListRepository amlWhiteListRepository;

    private final ExampleJdbc exampleJdbc;

    ExampleServiceImpl(UserRepository userRepository,
                       InfoRepository infoRepository,
                       AmlWhiteListRepository amlWhiteListRepository,
                       ExampleJdbc exampleJdbc) {
        this.userRepository = userRepository;
        this.infoRepository = infoRepository;
        this.amlWhiteListRepository = amlWhiteListRepository;
        this.exampleJdbc = exampleJdbc;
    }

    @Override
    public List<User> findUsers() {
        return new ArrayList<>(
                Arrays.asList(
                        new User("Gary", "M"),
                        new User("Mary", "F"),
                        new User("Cindy", "F"),
                        new User("Johnson", "M"),
                        new User("Fiona", "F")
                )
        );
    }

    @DS(value = "example2")
    public void findExample1Data() {
        List<UserEntity> list = userRepository.findAll();
        list.forEach(data -> {
            log.debug(data.toString());
        });
    }

    @DS(value = "example1")
    public void findExample2Data() {
        List<InfoEntity> list = infoRepository.findAll();
        list.forEach(data -> {
            log.debug(data.toString());
        });
    }

    @Override
    public void findDb2Data() {
        List<AmlWhiteListEntity> list = amlWhiteListRepository.findAll();
        list.forEach(data -> {
            log.debug(data.toString());
        });
    }

    @Override
    @DS(value = "example2")
    public void findByNativeSQL(String name) {
        ResourceEnum resource = ResourceEnum.SQL.getResource("FIND_USER_BY_NAME");
        Map<String, Object> params = new HashMap<>();
        params.put("name", name);
        UserEntity user = exampleJdbc.queryForBean(resource, params, UserEntity.class);
        log.debug(user.toString());
    }

    @Override
    public String testException() {
        throw new RuntimeException("拋出一個異常");
    }
}

File: example-kotlin/src/main/kotlin/com/example/service/impl/ExampleServiceImpl.kt
package com.example.service.impl

import com.example.jdbc.ExampleJdbc
import com.example.model.UserDetail
import com.example.model.UserMain
import com.example.repository.UserDetailRepository
import com.example.repository.UserMainRepository
import com.example.service.ExampleService
import com.zipe.base.annotation.DS
import com.zipe.enums.ResourceEnum
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ExampleServiceImpl : ExampleService {

    @Autowired
    lateinit var userMainRepository: UserMainRepository

    @Autowired
    lateinit var userDetailRepository: UserDetailRepository

    @Autowired
    lateinit var exampleJdbc: ExampleJdbc

    @DS
    override fun findUserByName(name: String) = userMainRepository.findUserMainByName(name)

    @DS("example")
    override fun findUserInfoByName(name: String): UserDetail? = userDetailRepository.findById(name).get()

    @DS
    override fun findUserByJdbc(name: String): List<UserMain>? {
        val resource = ResourceEnum.SQL.getResource("FIND_USER_MAIN_BY_NAME")
        val param = mutableMapOf<String, Any>("name" to name)
        return exampleJdbc.queryForList(resource, param, UserMain::class.java).toList()
    }

    @DS
    @Transactional
    override fun saveOrUpdateUser(userMain: UserMain) = userMainRepository.save(userMain)

    @DS("example")
    @Transactional
    override fun saveOrUpdateInfo(userDetail: UserDetail) = userDetailRepository.save(userDetail)

    @DS
    @Transactional
    override fun deleteUser(userMain: UserMain) = userMainRepository.delete(userMain)

    @DS("example")
    @Transactional
    override fun deleteInfo(userDetail: UserDetail) = userDetailRepository.delete(userDetail)
}
